### PR TITLE
Autotune flashinfer extend buckets at warmup

### PR DESCRIPTION
**Pull request by sshleifer**
## Problem

The warmup flashinfer autotune runs a **decode-shaped** dummy forward (`BaseRunner.warmup` → `_autotune_buffers()`, 1 token/req), so tuned buckets only cover token counts up to the decode batch size. Larger prefill/extend batches fall outside the tuned buckets at serving time and run flashinfer's slow default heuristic:

```
[AutoTuner]: No tuned config covers flashinfer::trtllm_fp4_block_scale_moe input_shapes=(torch.Size([16384, 7168]), ...);
falling back to runner=MoERunner tactic=-1. This shape is outside the tuning bucket range ...
```

With flashinfer **0.6.15.post1**, the default heuristic for `trtllm_fp4_block_scale_routed_moe` is **~30% slower than the tuned tactic at >=8192 tokens** on sm100 (GB300) — flashinfer-ai/flashinfer#4190. Microbench at Kimi-K2-architecture shapes (384 experts, top-8, hidden 7168, intermediate 512/rank at tp4), same GPU, µs/call:

| tokens | 0.6.13 heuristic | 0.6.15.post1 heuristic | 0.6.13 autotuned | 0.6.15.post1 autotuned |
|---|---|---|---|---|
| 1024 | 567 | **443 (−22%)** | 550 | **402 (−27%)** |
| 4096 | 607 | 593 | 554 | 545 |
| 8192 | 840 | **1074 (+28%)** | 839 | 825 |
| 16384 | 1319 | **1713 (+30%)** | 1313 | 1304 |

End-to-end: **−7…−9% extend throughput** on TP-only NVFP4 MoE serving at >=8k-token extend batches (dp-attention configs stay under the bucket per rank and are unaffected).

## Fix

One **EXTEND-shaped** autotune dummy forward at warmup:

- **Sized to `max_prefill_tokens`** (÷ `dp_size` under dp-attention) — the scheduler's per-forward-batch extend-token budget, i.e. the largest shape the kernels see; tuning there covers every smaller bucket. `chunked_prefill_size` would under-cover (it caps per-request chunks; chunks batch together up to `max_prefill_tokens`). Deliberately unbounded, no new knob: serving must survive a real forward at that exact shape, buffers are freed after the pass, and large-context configs are the ones whose workload lives in the big buckets. Disable via `--disable_flashinfer_autotune`.
- **Packed within the req pool**: the dummy batch is at most `req_to_token_pool.size` seqs × N tokens (e.g. 1024×16) via `_dummy_run(extend_num_tokens_per_req=...)` — per-seq attention metadata (e.g. triton `kv_indptr`) is sized to the req pool, so num_tokens 1-token seqs would crash configs with `max_prefill_tokens > max_num_seqs`. Also matches real chunked-prefill shapes. Autotune buckets key on num_tokens, so tuning results are unchanged.
- The dummy buffer set skips the `(num_tokens, vocab)` fp32 logits buffer (>10 GB at 16k tokens) since autotune runs with `skip_logits`.
- Speculative-decoding runners are skipped for now (`_dummy_run` forces TARGET_VERIFY shapes there); spec configs are a follow-up.

## Verification (GB300, Kimi-K2-arch NVFP4 MoE, tp4, flashinfer 0.6.15.post1, `batch_full` bench)

prefill extend tok/s (bs=1, n=3, IQR≈0):

| input_len | 0.6.13 | 0.6.15.post1 unfixed | 0.6.15.post1 + this PR |
|---|---|---|---|
| 4096 | 47,270 | 47,357 | 48,473 |
| 8192 | **53,155** | **48,957 (−7.9%)** | **53,401 (+0.7% vs 0.6.13)** |
| 16384 | **51,946** | **47,877 (−7.8%)** | **51,349 (+0.4%)** |

decode_long@8192 +9.6%, saturated-throughput +9.3% (ttft −17%) vs unfixed; numerics unchanged (wikitext / long-decode ppl identical to 4 digits across versions). The packed dummy (1024×16) validated end-to-end on a config with `max_num_seqs=1024, max_prefill_tokens=16384` that crashed with the unpacked shape.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #31142267174](https://github.com/sgl-project/sglang/actions/runs/31142267174)
Latest PR Test (Extra): :x: [Run #31142267141](https://github.com/sgl-project/sglang/actions/runs/31142267141)

### CODE REVIEW

**Comment by sshleifer** [comment]
**Design note — why a dummy forward in `BaseRunner.warmup()` rather than wrapping the existing `--warmups prefill_shapes` sweep in the autotune context:**

1. **Process boundary.** `prefill_shapes` runs in the entrypoint/tokenizer-manager process; `flashinfer.autotuner.autotune(True)` is a context manager around forward execution inside each scheduler process. Wiring it up would need new IPC (enter/exit-tuning-mode messages to every TP/DP rank bracketing the sweep) — more machinery than this PR.

2. **Tuning during live serving is unsafe-by-construction.** `--warmups` run after the server starts accepting traffic (that's their design point — they exercise the real request path). A real request arriving while ranks are in tuning mode gets served by profiling runs (tactic sweeps, extra syncs, autotuner state mutation); the autotuner is not built for concurrent use with overlap-scheduled serving. Runner warmup is the one moment execution is guaranteed quiescent.

3. **Ordering vs prefill CUDA-graph capture.** `--warmups` run after graph capture, and capture bakes whatever tactic the cache holds at that moment — tuning afterwards can't fix already-captured extend graphs. In the runner-warmup slot the pass runs before capture (`capture()` calls `warmup()` first in both graph runners), so captured prefill graphs get the tuned tactics too. Verified with `enable_prefill_cuda_graph=True`.

4. **Default-on and cheap.** `--warmups prefill_shapes` is opt-in and costs ~20 generate round-trips; deployments that don't pass the flag would keep the regression silently. This pass is one extra forward inside the existing run-once autotune moment, on by default, ~seconds, with an env kill-switch.

Longer term it could be worth unifying "shapes we JIT-warm" (`prefill_shapes`) and "shapes we tune" (this pass) into one declaration consumed by both — happy to follow up if maintainers think that's the right direction.

**Comment by sshleifer** [comment]
Updated: dropped the `SGLANG_FLASHINFER_AUTOTUNE_MAX_EXTEND_TOKENS` env var — the pass now sizes directly from `max_prefill_tokens` (the scheduler's per-forward-batch extend-token budget, so it covers every real shape; `chunked_prefill_size` would under-cover since chunks batch together). Off-switch is the existing `--disable_flashinfer_autotune`.

File: python/sglang/srt/model_executor/runner/base_runner.py
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/base_runner.py:65]: Can we move all the flashinfer autotune related functions to `python/sglang/srt/model_executor/runner/flashinfer_autotune.py`, including `_flashinfer_autotune`, `_dummy_run`,  `_allocate_decode_buffers`, `_maybe_flashinfer_autotune_extend`...

File: python/sglang/srt/model_executor/runner/base_runner.py [hunk 1]
```diff
@@ -294,7 +303,75 @@ def forward_fn():
...
+            # scale.
+            num_tokens //= mr.server_args.dp_size
+        if num_tokens <= (decode_num_tokens or 0):
+            return  # decode-shaped autotune already covered these buckets
+        if not mr.is_generation or mr.spec_algorithm.is_speculative():
+            # _dummy_run forces TARGET_VERIFY shapes for speculative runners;
+            # extend-bucket autotune for spec configs is a follow-up.
+            return
+
+        num_tokens_per_req = -(-num_tokens // mr.req_to_token_pool.size)
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/base_runner.py]: It's a little weird to do double negative here, can we remove the -(-

File: python/sglang/srt/model_executor/runner/base_runner.py [hunk 2]
```diff
@@ -294,7 +303,75 @@ def forward_fn():
...
+        far slower than the tuned tactic (e.g. trtllm-gen fp4 MoE is ~30%
+        slower untuned at >=8k tokens on sm100). One extra forward at the
+        largest per-rank extend token count tunes all buckets up to it.
+        """
+        mr = self.model_runner
+        # max_prefill_tokens bounds total extend tokens per forward batch, so
+        # tuning there covers every real extend shape (chunked_prefill_size
+        # only caps per-request chunks). Unbounded on purpose: serving runs
+        # this exact shape anyway; the cost is startup tuning time.
+        num_tokens = mr.server_args.max_prefill_tokens
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/base_runner.py]: Do we need to autotune on more `num_tokens`? Will only autotuning on `max_prefill_tokens` not enough?

Comment (paulzhang-tm) [reply python/sglang/srt/model_executor/runner/base_runner.py]: Max M is likely sufficient given gemm characteristics?

File: python/sglang/srt/model_executor/runner/base_runner.py
Comment (sshleifer) [reply python/sglang/srt/model_executor/runner/base_runner.py:65]: Done — the extend pass now lives in `runner/flashinfer_autotune.py` (`maybe_flashinfer_autotune_extend(runner, ...)`, matching the existing `maybe_flashinfer_autotune_speculative_draft` shape); `base_runner.py` keeps only the warmup call site.

I left `_dummy_run` / `_alloc_dummy_decode_buffers` in `base_runner.py`: they predate this PR and are shared with the non-autotune warmup and graph-capture paths, so moving them is a bigger containment change than this PR should carry. Happy to do that as a follow-up if you'd like.

-Claude (approved by sam)

File: python/sglang/srt/model_executor/runner/base_runner.py (hunk 1 above)
Comment (sshleifer) [reply python/sglang/srt/model_executor/runner/base_runner.py]: Gone — the remaining ceil-division (only the req-pool packing branch still needs one) is now the explicit `(num_tokens + pool_size - 1) // pool_size` form.

-Claude (approved by sam)

File: python/sglang/srt/model_executor/runner/base_runner.py (hunk 2 above)
Comment (sshleifer) [reply python/sglang/srt/model_executor/runner/base_runner.py]: Agreed — and that's what the patch does: a single extra pass at `max_prefill_tokens` (no per-bucket loop), since one forward at the largest M tunes all buckets up to it, the same way the decode-shaped pass already covers everything up to the decode batch size.

One caveat worth recording since it's not obvious: "covered" isn't automatically "well-tuned." We found on an internal GB300 serving benchmark (details in the PR-level comment) that tactics profiled under different seq compositions of the *same* token bucket differ by 15-20% in serving effect — a pool-packed `1024 x 16` dummy and a chunk-shaped `1 x 16384` dummy both tuned materially worse tactics than `16384 x 1` with identical bucket coverage. So the single max-M pass deliberately uses many 1-token seqs; the composition is load-bearing, not incidental.

-Claude (approved by sam)

File: python/sglang/srt/model_executor/runner/base_runner.py [hunk 3]
```diff
@@ -294,7 +302,60 @@ def forward_fn():
...
+        mr = self.model_runner
+        num_tokens = min(
+            envs.SGLANG_FLASHINFER_AUTOTUNE_MAX_EXTEND_TOKENS.get(),
+            mr.server_args.max_prefill_tokens or 0,
+        )
+        if require_mlp_tp_gather(mr.server_args):
+            # dp-attention splits an extend batch across dp ranks; tune the
+            # per-rank share so the dummy's gathered buffers stay at serving
+            # scale.
+            num_tokens //= mr.server_args.dp_size
```
[the remark this answers is not in the document]
Comment (sshleifer) [reply python/sglang/srt/model_executor/runner/base_runner.py]: Confirmed, and fixed — the division is removed. Two code facts check out:

1. `max_prefill_tokens` is enforced by `PrefillAdder`, which is per-scheduler — i.e. per DP rank. Each rank independently forms local extend batches up to the full budget, so dividing the tuning size by `dp_size` leaves the buckets ranks actually execute untuned (at DP4: tuned 4096, executed 16384).
2. It also under-tuned the gathered side: under mlp-tp-gather the MLP/MoE's M is `sum(global_num_tokens)` (`dp_attention.py`), and since warmup runs on all ranks simultaneously, the gathered dummy M is `dp_size x per-rank-dummy-size`. With the division that gather peaked at `max_prefill_tokens`; without it, it reaches `dp_size x max_prefill_tokens` — the true worst-case serving gather — so the undivided pass now covers both the per-rank and gathered buckets.

The cost is a transiently larger gathered dummy buffer during warmup; it's a shape serving must survive anyway, and the buffers are freed before graph capture measures free memory.

-Claude (approved by sam)

File: python/sglang/srt/model_executor/runner/base_runner.py [hunk 4]
```diff
@@ -294,7 +302,60 @@ def forward_fn():
...
+            if (c := mr.canary_manager) is not None
+            else empty_context()
+        )
+
+        def forward_fn():
+            self._dummy_run(
+                batch_size=num_tokens,
+                buffers=buffers,
+                run_ctx=canary_run_ctx,
+                forward_mode_override=ForwardMode.EXTEND,
```
[the remark this answers is not in the document]
Comment (sshleifer) [reply python/sglang/srt/model_executor/runner/base_runner.py]: Good catch — the forced-EXTEND dummy leaves `mm_inputs=None`, and llava-style prefill paths iterate it. This revision skips the extend pass for multimodal models (`model_config.is_multimodal`), alongside the existing skip for speculative configs. Proper placeholder mm metadata so multimodal models also get extend-bucket tuning is a reasonable follow-up, but I'd rather not block the common case on it.

-Claude (approved by sam)

**Comment by sshleifer** [comment]
Pushed an update addressing the review round, plus one change from production experience with this patch internally that reviewers should see first:

**The dummy's seq composition affects tuned-tactic quality, not just bucket coverage.** An earlier internal revision of this patch ceil-packed the dummy into the req pool (`1024 seqs x 16 tokens` instead of `16384 x 1` on the affected config). Same token bucket — but on an internal GB300 serving benchmark (Kimi-architecture NVFP4 MoE, PD-disaggregated), cold-start saturated peak decode throughput dropped 15-20% below an 8-day baseline floor (2550-2796 tok/s vs a 3136-4230 band over 22 runs), fully recovered on revert, and a chunk-shaped `1 x 16384` dummy was just as bad (2697). Only many-1-token-seqs tuned the good tactics. Client-visible ttft/tpot were unchanged — the damage is confined to tactic quality under load bursts, which makes it nasty to catch.

So this revision:
1. runs the dummy as 1-token seqs everywhere, except backends that preallocate per-seq extend metadata at req-pool size (triton's `kv_indptr` crashes on `batch_size > pool` at startup) — expressed as an internal `AttentionBackend.extend_dummy_seqs_capped_by_req_pool` capability, OR-composed in `HybridAttnBackend`;
2. tunes a single pass at the full per-rank `max_prefill_tokens` (dp_size division removed — see inline reply);
3. skips multimodal models (see inline reply);
4. moves the pass into `runner/flashinfer_autotune.py`.

-Claude (approved by sam)

**Comment by sshleifer** [comment]
Pushed one more hardening commit: the pool-cap capability is now propagated through the composite wrappers (`TboAttnBackend`, `HybridLinearAttnBackend`, `MiniMaxHybridAttnBackend`, mirroring the existing `needs_cpu_seq_lens` delegation), and `HybridAttnBackend` derives it from the prefill backend alone since `_select_backend` routes EXTEND there unconditionally. Without this, a triton backend wrapped by TBO or a hybrid-linear model would have read the base-class default and reproduced the `kv_indptr` startup crash.

-Claude (approved by sam)

**Review by Fridge003 — APPROVED** [review]

**Comment by sshleifer** [comment]
Heads-up on a post-approval change: the last two CI rounds showed the extend-shaped warmup forward costs more memory headroom than some configs have — `base-b-test-1-gpu-small (3)` (sm120, 31 GiB) OOMs inside the tuned fp4 GEMM mid-pass, and `test_nvidia_nemotron_3_nano` (2×80 GiB) OOMs at serving time after warmup. Both shards pass on a sibling PR without this change, so it's this PR's headroom cost.

I've made the pass opt-in (`SGLANG_FLASHINFER_AUTOTUNE_EXTEND`, default off) and fail-open on `OutOfMemoryError` (log + skip, untuned extend buckets as before). The req-pool-cap fixes on the indptr-bounded backends stay. Happy to flip the default in a follow-up once the pass sizes itself against free memory.
